helm-imenu.el

```diff
@@ -270,22 +298,29 @@ only '((foo . bar)) is needed."
 A mode is similar as current if it is the same, it is derived i.e `derived-mode-p'
 or it have an association in `helm-imenu-all-buffer-assoc'."
   (interactive)
-  (unless helm-source-imenu-all
-    (setq helm-source-imenu-all
-          (helm-make-source "Imenu in all buffers" 'helm-imenu-source
-            :init (lambda ()
-                    ;; Use a cache to avoid repeatedly sending
-                    ;; progress-reporter message when updating
-                    ;; (Issue #1704).
-                    (setq helm-imenu--in-all-buffers-cache
-                          (helm-imenu-candidates-in-all-buffers)))
-            :candidates 'helm-imenu--in-all-buffers-cache
-            :fuzzy-match helm-imenu-fuzzy-match)))
+  (unless helm-imenu-in-all-buffers-separate-sources
+    (unless helm-source-imenu-all
+      (setq helm-source-imenu-all
+            (helm-make-source "Imenu in all buffers" 'helm-imenu-source
+              :init (lambda ()
+                      ;; Use a cache to avoid repeatedly sending
+                      ;; progress-reporter message when updating
+                      ;; (Issue #1704).
+                      (setq helm-imenu--in-all-buffers-cache
+                            (helm-imenu-candidates-in-all-buffers)))
+              :candidates 'helm-imenu--in-all-buffers-cache
+              :fuzzy-match helm-imenu-fuzzy-match))))
   (let ((imenu-auto-rescan t)
         (str (thing-at-point 'symbol))
         (helm-execute-action-at-once-if-one
-         helm-imenu-execute-action-at-once-if-one))
-    (helm :sources 'helm-source-imenu-all
+         helm-imenu-execute-action-at-once-if-one)
+        (helm--maybe-use-default-as-input
+         (not (null (memq 'helm-source-imenu-all
```

Comment:
`memq` is not returning a boolean so it is cleaner to return a boolean, though yes it would work also of course with a list instead of `t`.